File: src/main/java/com/handson/RandomSequenceGenerator.java
package com.handson;


import rx.Observable;

import java.util.Random;
import java.util.concurrent.TimeUnit;

public class RandomSequenceGenerator {

    private final double min;
    private final double max;
    private final Random random;

    public RandomSequenceGenerator(double min, double max) {
        this.min = min;
        this.max = max;
        this.random = new Random();
    }

    public Observable<Double> create(long interval, TimeUnit timeUnit) {
        double init = (min + max) /2;
        return Observable
                .interval(interval, timeUnit)
                .scan(init, (previous, i) -> computeNextNumber(previous));
    }

    public double computeNextNumber(double previous) {
        double range = (max - min) / 100;
        double scaled = (random.nextDouble() - 0.5) * range;
        double shifted = previous + scaled;
        if (shifted < min || shifted > max) {
            shifted = previous - scaled;
        }
        return shifted;
    }

}

File: CSharp/NetCore/MarketData/src/MarketDataCommon/Infra/RandomSequenceGenerator.cs
using System;
using System.Reactive.Linq;

namespace MarketDataCommon.Infra
{
    public class RandomSequenceGenerator
    {
        private readonly double _min;
        private readonly double _max;
        private readonly Random _random;
        private double _bias = 0.3;

        public RandomSequenceGenerator(double min, double max)
        {
            _min = min;
            _max = max;
            _random = new Random();
        }

        public IObservable<double> Create(TimeSpan timeSpan)
        {
            double init = (_min + _max) / 2;
            return Observable.Interval(timeSpan)
                             .Scan(init, (previous, i) => ComputeNextNumber(previous));
        }

        public IObservable<int> CreateIntegerSequence(TimeSpan timeSpan)
        {
            double range = _max - _min;
            return Observable.Interval(timeSpan)
                             .Select(i => (int)(_random.NextDouble() * range + _min));

        }

        private double ComputeNextNumber(double previous)
        {
            double range = (_max - _min) / 20;
            double scaled = (_random.NextDouble() - 0.5 + _bias) * range;
            double shifted = previous + scaled;

            if (shifted < _min || shifted > _max)
            {
                shifted = previous - scaled;
                _bias = -_bias;
            }

            shifted = (double)((long)Math.Round(shifted * 10000)) / 10000;

            return shifted;
        }
    }
}
